Update the text to fit in the module change; provide usage examples (#40)
boucadair/policy-based-network-acl: draft-ma-opsawg-schedule-yang.md

#  Module Overview

   The "ietf-schedule" module defines the following "grouping" statement:

   * period-of-time
   * recurrence
   * recurrence-with-date-times
   * icalendar-recurrence

   {{schedule-tree}} provides an overview of the tree structure {{?RFC8340}} of
   the "ietf-schedule" module in terms of its groupings.

~~~~
{::include ./yang/ietf-schedule-tree.txt}
~~~~
{: #schedule-tree title="Schedule Tree Structure" artwork-align="center"}

   Each of these groupings are presented in the following subsections. Examples
   are provided in {{usage}}.

## The "period-of-time" Grouping

   The "period-of-time" grouping allows a time period to be represented using
   either a start ("period-start") and end date and time ("period-end"), or a
   start ("period-start") and a positive time duration ("duration"). For the first
   format, the start of the period MUST be before the end of the period

>    {{Section 3.3.9 of ?RFC5545}}.

Author: I am unsure about this, should it be a normative reference?

>    {{Section 3.3.9 of ?RFC5545}}.

Reviewer: Yes as many rules are defined there. Thanks.

>    {{Section 3.3.9 of ?RFC5545}}.

Author: ok.

## The "recurrence" Grouping

  The "recurrence" grouping specifies a simple recurrence rule, the definition

>   conforms to part of the "recurrence rule" properties in {{Section 3.3.10 of ?RFC5545}}.

Reviewer: Same reply, then :-)

--- back

# Examples of Format Representation {#usage}

   This section provides some examples to illustrate the use of the
   period and recurrence formats defined as YANG groupings. Note that "grouping"

>    does not define any nodes in the schema tree, the examples illustrated are

Reviewer: ```suggestion
   does not define any data nodes in the schema tree, the examples illustrated are
```

   just for the ease of understanding. Only the message body is provided with
   JSON used for encoding {{?RFC7951}}.

## The "period-of-time" Grouping

>    The example of a period that starts at 08:00:00 UTC, on January 1, 2023 and ends at 18:00:00 UTC

Reviewer: ```suggestion
   The example of a period that starts at 08:00:00 UTC, on January 1, 2025 and ends at 18:00:00 UTC
```

   on December 31, 2025 is encoded as follows:

~~~~
{

>   "period-start": "2023-01-01T08:00:00Z",

Reviewer: ```suggestion
  "period-start": "2025-01-01T08:00:00Z",
```

>   "period-end": "2025-12-01T18:00:00Z"

Reviewer: ```suggestion
  "period-end": "2027-12-01T18:00:00Z"
```

~~~~
{

>   "period-start": "2023-01-01T08:00:00Z",

Reviewer: ```suggestion
  "period-start": "2025-01-01T08:00:00Z",
```

  "duration": "P15DT05:20:00"
}
~~~~

   An example of a period that starts at 2:00 A.M. in Los Angeles on November 19,

>    2023 and lasts 20 weeks:

Reviewer: ```suggestion
   2025 and lasts 20 weeks:
```

~~~~
{

>   "period-start": "2023-11-19T02:00:00",

Reviewer: ```suggestion
  "period-start": "2025-11-19T02:00:00",
```

  "time-zone-identifier": "America/Los_Angeles",
  "duration": "P20W"
}
~~~~

## The "recurrence" Grouping

>    The following snippet can be used to indicate a every other day recurrence for

Reviewer: ```suggestion
   The following snippet can be used to indicate a daily recurrence for
```

Author: There is a difference between "every other day" and "daily", by "daily", the interval value is 1, while by "every other day", we mean “every two days”, the interval value is 2.

>    10 occurrences, starting at 3 p.m. on December 1, 2023 in the Eastern United States time zone:

Reviewer: ```suggestion
   10 occurrences, starting at 3 p.m. on December 1, 2025 in the Eastern United States time zone:
```

~~~~
{
  "recurrence-first": {

>     "date-time-start": "2023-11-01T15:00:00",

Reviewer: ```suggestion
    "date-time-start": "2025-11-01T15:00:00",
```

    "time-zone-identifier": "America/New_York"
  },
  "frequency": "ietf-schedule:daily",
  "interval": 2,
  "count": 10
}
~~~~

   An example of an anniversary that will occur annually, from 1997-11-25, until 2050-11-25:

~~~~
{
  "recurrence-first": {
    "date-time-start": "1979-11-25"
  },
  "frequency": "ietf-schedule:yearly",
  "until": "2050-11-25"
}
~~~~

## The "recurrence-with-date-times" Grouping

>    The following snippet can be used to indicate a recurrence that occur every 2

Reviewer: ```suggestion
   The following snippet can be used to indicate a recurrence that occurs every 2
```

>    hours from 9:00 AM to 5:00 PM, and 6PM UTC time on 2023-12-01:

Reviewer: ```suggestion
   hours from 9:00 AM to 5:00 PM, and 6PM UTC time on 2025-12-01:
```

~~~~
{
  "recurrence-first": {

>     "date-time-start": "2023-12-01T09:00:00Z"

Reviewer: ```suggestion
    "date-time-start": "2025-12-01T09:00:00Z"
```

  },
  "frequency": "ietf-schedule:hourly",
  "interval": 2,

>   "until": "2023-12-01T17:00:00Z",

Reviewer: ```suggestion
  "until": "2025-12-01T17:00:00Z",
```

>   "date-times": ["2023-12-01T18:00:00Z"]

Reviewer: ```suggestion
  "date-times": ["2025-12-01T18:00:00Z"]
```

}
~~~~

>    The following snippet can be used to indicate a recurrence that occur every

Reviewer: ```suggestion
   The following snippet can be used to indicate a recurrence that occurs every
```

   30 minutes and last for 15 minutes from 9:00 AM to 5:00 PM, and extra two occurrences

>    at 6:00 PM and 6:30 PM with each lasting for 20 minutes on 2023-12-01:

Reviewer: ```suggestion
   at 6:00 PM and 6:30 PM with each lasting for 20 minutes on 2025-12-01:
```

~~~~
{
  "recurrence-first": {

>     "date-time-start": "2023-12-01T09:00:00Z",

Reviewer: ```suggestion
    "date-time-start": "2025-12-01T09:00:00Z",
```

    "duration": "PT00:00:15"
  },
  "frequency": "ietf-schedule:minutely",
  "interval": 30,

>   "until": "2023-12-01T17:00:00Z",

Reviewer: ```suggestion
  "until": "2025-12-01T17:00:00Z",
```

  "period": [
    {

>       "period-start": "2023-12-01T18:00:00Z",

Reviewer: ```suggestion
      "period-start": "2025-12-01T18:00:00Z",
```

      "duration": "PT00:00:20"
    },
    {

>       "period-start": "2023-12-01T18:30:00Z",

Reviewer: ```suggestion
      "period-start": "2025-12-01T18:30:00Z",
```

      "duration": "PT00:00:20"
    }
   ]
}
~~~~

## The "icalendar-recurrence" Grouping

>    The following snippet can be used to indicate 10 occurrences that occur at

Reviewer: ```suggestion
   The following snippet indicates 10 occurrences that occur at
```

   8:00 AM (EST), every last Saturday of the month starting in January 2024:

~~~~
{
  "recurrence-first": {
    "date-time-start": "2024-01-27T08:00:00",
    "time-zone-identifier": "America/New_York"
  },
  "frequency": "ietf-schedule:monthly",
  "count": 10,
  "byday": [
    {
      "direction": [-1],
      "weekday": "saturday"
    }
  ]
}
~~~~

>    The following indicates the example of a recurrence that occurs on the last

Reviewer: ```suggestion
   The following is an example of a recurrence that occurs on the last
```

   workday of the month until December 25, 2024, from January 1, 2024:

~~~~
{
  "recurrence-first": {
  "date-time-start": "2024-01-01"
  },
  "frequency": "ietf-schedule:monthly",
  "until": "2024-12-25",
  "byday": [
    { "weekday": "monday"},
    { "weekday": "tuesday"},
    { "weekday": "wednesday"},
    { "weekday": "thursday"},
    { "weekday": "friday"}
  ],
  "bysetpos": [-1]
}
~~~~

  The following snippet can be used to indicate a recurrence that occur every 20
  minutes from 9:00 AM to 4:40 PM (UTC), with the occurrence starting at 10:20 AM

>   being excluded on 2023-12-01:

Reviewer: ```suggestion
  being excluded on 2025-12-01:
```

~~~~
{
  "recurrence-first": {

>   "date-time-start": "2023-12-01T09:00:00Z"

Reviewer: ```suggestion
    "date-time-start": "2023-12-01T09:00:00Z"
```

>   "date-time-start": "2023-12-01T09:00:00Z"

Reviewer: ```suggestion
  "date-time-start": "2025-12-01T09:00:00Z"
```

  },

>   "until": "2023-12-01T16:40:00Z",

Reviewer: ```suggestion
  "until": "2025-12-01T16:40:00Z",
```

  "frequency": "ietf-schedule:minutely",
  "byminute": [0, 20, 40],
  "byhour": [9, 10, 11, 12, 13, 14, 15, 16],

>   "exception-dates": ["2023-12-01T10:20:00Z"]

Reviewer: ```suggestion
  "exception-dates": ["2025-12-01T10:20:00Z"]
```

   every Friday at 8:00 AM). The following example module which "uses" the "icalendar-recurrence"
   grouping from "ietf-schedule" module shows how a scheduled task could be defined
   with different features used for options.

~~~~
{::include ./yang/example-scheduled-backup.yang}
~~~~

## Example: Schedule Network Properties to Change Based on Date and Time {#augments}

   Network properties may change over a specific period of time or based on a
   recurrence rule, e.g., {{?I-D.ietf-tvr-use-cases}}.
   The following example module which augments the "recurrence-with-date-times"
   grouping from "ietf-schedule" module shows how a scheduled based attribute
   could be defined.

~~~~
{::include ./yang/example-scheduled-link-bandwidth.yang}
~~~~

  Following is a configuration example which indicates a link's bandwidth is
  scheduled between 2023/12/01 0:00 UTC to the end of 2023/12/31 with a daily
  schedule. In each day, the bandwidth value is scheduled to be 500Kbps between
  1:00 AM to 6:00 AM and 800Kbps between 10:00 PM to 11:00 PM. The bandwidth
  value that's not covered by the period above is 1000Kbps by default.

~~~~

> {::include ./examples/example-scheduled-link-bandwidth.xml}

Reviewer: ```suggestion
{::include-fold ./examples/example-scheduled-link-bandwidth.xml}
```
